ffi/src/lib.rs

```diff
@@ -0,0 +1,180 @@
+        size: u32,
+        rate_max: u32,
+        rate_seconds: u32,
+    ) -> Result<bool> {
+        let prefix = prefix.to_str()?;
+
+        /* re-connect if necessary */
+        if let None = self.client {
+            self.client = Some(Client::connect(String::from(self.url.clone()))?)
+        };
+
+        if let Some(client) = &self.client {
+            let now = SystemTime::now()
+                .duration_since(SystemTime::UNIX_EPOCH)?
+                .as_secs();
+
+            let key = format!("{}{}", prefix, now);
+            match client.add(&key, 0, rate_seconds + 1) {
+                Ok(()) => (),
+                Err(e) => match e {
+                    CommandError(e) if e == KeyExists => (),
+                    e => bail!(e),
+                },
+            };
+
+            let keys: Vec<String> = (0..rate_seconds)
+                .map(|n| format!("{}{}", prefix, n))
+                .collect();
+            let keys: Vec<&str> = keys.iter().map(|s| &**s).collect();
+            let tokens: HashMap<String, u32> = client.gets(&keys)?;
+            let (value, cas_id) = match client.get(&key)? {
+                Some((value, _, cas)) => (
+                    match str::from_utf8(&value) {
+                        Ok(value) => match value.parse::<u32>() {
+                            Ok(value) => value,
+                            Err(_) => 0,
+                        },
+                        Err(_) => 0,
+                    },
+                    cas,
+                ),
+                None => (0, None),
+            };
+
+            let mut sum = value + size;
+            for (_, v) in tokens {
+                sum += v;
+            }
+
+            if sum > rate_max {
+                return Ok(true);
+            }
+
+            if let Some(cas_id) = cas_id {
+                client.cas(&key, value + size, rate_seconds + 1, cas_id)?;
+            }
+
+            Ok(false)
+        } else {
+            // TODO
+            // shouldn't get in here?
+            // but should probably error?
+            Ok(false)
+        }
+    }
+}
+
+thread_local!(
+  static COUNTER: Cell<u64> = Cell::new(1);
+  static STORE: RefCell<HashMap<u64, Rl>> = RefCell::new(HashMap::new())
+);
+
+#[no_mangle]
+pub extern "C" fn rl_new(url: *const i8) -> u64 {
+    let url = unsafe { CStr::from_ptr(url) };
+    let rl = match Rl::new(url) {
+        Ok(rl) => rl,
+        Err(_) => return 0,
+    };
+
+    let index = COUNTER.with(|it| {
+        let index = it.get();
+        it.set(index + 1);
+        index
+    });
+
+    STORE.with(|it| {
+        let mut it = it.borrow_mut();
+        it.insert(index, rl);
+    });
+
+    index
+}
+
+#[no_mangle]
+pub extern "C" fn rl__rate_limit(
+    index: u64,
+    prefix: *const i8,
+    size: u32,
+    rate_max: u32,
+    rate_seconds: u32,
+) -> i32 {
+    let prefix = unsafe { CStr::from_ptr(prefix) };
+
+    STORE.with(|it| {
+        let mut it = it.borrow_mut();
```

Comment:
Novel way to avoid a `unsafe` blocks for every constructor/method/destructor.  [As can be seen in this example](https://metacpan.org/pod/FFI::Platypus::Lang::Rust#Rust-Source8).  Instead of passing opaque pointers back and forth from Rust to Perl, we store the objects in a Rust HashMap with an incrementing integer key.  Yes, there is some overhead to doing the lookup, but for relatively small number of instances (I expect there to in practice usually only be one $rl object), this shouldn't be a problem.

If a bad argument gets passed in, say index 42 when there is no corresponding Rl object, we can just return false, issue an error and keep going without crashing.  If we were passing pointers back and forth there isn't actually a way to know if the pointer is bad (other than NULL pointer error).  So this should in theory be safer.

[rustc-hash](https://github.com/rust-lang/rustc-hash) may be more performant for larger object sets, but I haven't done any benchmarking yet.  [Acme::Color::Rust](https://metacpan.org/pod/Acme::Color::Rust) was created to experiment with this technique and might be a good place to do such benchmarking.